### PR TITLE
Fix for infinite loop bug that can crash chrome. 

### PR DESCRIPTION
**Pull request by vladloom**
**Problem**
Given two concurrent dispatches into the redux store, across different threads, we can fall into an infinite loop of saving competing states. `_send2Storage` saves State A, `subscribe` responds to a save of State B, saves State B, and then we go back to A - until Chrome crashes. 

I've set up a repo where the problem can be replicated:
https://github.com/vladloom/chrome-extension-reduxed-storage

Here's a Demo of the crash in action:
https://www.loom.com/share/35a654ae83b540d4953c4aa89f1485c4

**Error condition**: 
newTime = false
isEqual = false

**Before**:

```
newTime && isEqual(newState, state) ||
  this._send2Storage();
```

| newTime | isEqual(newState, state)       |                                            |.                               |
|----------|-----------------------------|--------------------------|-------------------|
| false        | false                                        | this._send2Storage();      | Error condition      |
| false        | true                                        | this._send2Storage();      | Already equal —> no need to write |
| true.        | false                                        | this._send2Storage();      |                                |
| true         |  true                                        | no-op                                 |                                |
 

**After**:
```
  if(newTime && !isEqual(newState, state)) {
    this._send2Storage();
  }
```

| newTime | isEqual(newState, state)       |                                            |                        
|----------|-----------------------------|--------------------------|
| false        | false                                        | no-op                                 | 
| false        | true                                        | no-op                                 |
| true         | false                                        | this._send2Storage();      |
| true         |  true                                        | no-op                                 |

### CODE REVIEW

**Comment by hindmost** [comment]
Thanks for reporting this issue. It seems I'd overdone with optimization)

File: src/ReduxedStorage.ts [hunk 1]
```diff
@@ -79,7 +79,7 @@ export default class ReduxedStorage<
         mergeOrReplace(this.state, state) : mergeOrReplace(state, this.state);
       !newTime && isEqual(newState, this.state) ||
         ( this._setState(newState, timestamp), this._renewStore() );
-      newTime && isEqual(newState, state) ||
+      if (newTime && !isEqual(newState, state))
```
Comment (vhmth) [inline src/ReduxedStorage.ts:82]: Does the current logic preclude sending to storage when the new state is not equal? If so, does that mean the current logic is wrong?

**Comment by hindmost** [comment]
It seems I approved this PR too soon.
@vladloom, you were right that the `_send2Storage` condition was wrong. However your fix isn't quite suitable. The point is that an old timestamp may be a reason for `_send2Storage` too, although with proviso that the state is changed. I mean, if the store receives a state change with timestamp older than the current one (a rare but possible case) and the new and current states are unequal, `_send2Storage` has to be called too.
So the full condition:

```
newTime && !isEqual(newState, state) || !newTime && !isEqual(newState, state)
```

which is boiled down to:

```
!isEqual(newState, state)
```

Furthermore, it'd be better to place `_send2Storage` call with the condition into `if` statement for `_setState/_renewStore` call.
So the code after refactoring:

```
if (newTime || !isEqual(newState, this.state)) {
  this._setState(newState, timestamp);
  this._renewStore();
  if (!isEqual(newState, state)) {
    this._send2Storage();
  }
}
```

I'll include it into a new release soon.

File: src/ReduxedStorage.ts (hunk 1 above)
Comment (hindmost) [reply src/ReduxedStorage.ts:82]: Actually, it **does** sending when the new state is not equal

Comment (vhmth) [reply src/ReduxedStorage.ts:82]: https://www.loom.com/share/c1ddd0bd78144160aaefdc224cbd6d4e

Comment (hindmost) [reply src/ReduxedStorage.ts:82]: Well, no matter old or new code/logic, the qualifying condition is the same - inequality (aside from `newTime` condition). It seems you haven't noticed OR operator in the old code which negates the previous expression. In other words:
```
newTime && isEqual(newState, state) || something
```
is equivalent to:
```
if (!newTime || !isEqual(newState, state)) something
```

Comment (vhmth) [reply src/ReduxedStorage.ts:82]: I think we might be talking about 2 different things, but I do understand what you're saying now!

**Pull request merged**